draft-ietf-core-problem-details.md

```diff
@@ -13,224 +17,297 @@ stand_alone: yes
+     |                    |
 ~~~
-{: #fig-problem-details artwork-align="center" title="Problem Details"}
+{: #fig-problem-details artwork-align="center"
+   title="Problem Details: Example with CoAP"}
 
-The vocabulary presented is largely inspired by the Problem Details for HTTP APIs defined in {{?RFC7807}}.
+The framework presented is largely inspired by the Problem Details for HTTP APIs defined in {{RFC7807}}.
 
 ## Requirements Language
 
-{::boilerplate bcp14}
+{::boilerplate bcp14-tagged}
 
 # Basic Problem Details
 
-A CoRE Problem Details is a CoRAL document with the following elements:
+A Concise Problem Details data item is a CBOR data item with the following
+structure (notated in CDDL {{!RFC8610}}, using 65535 in place of a tag number to
+be defined for the type of problem details):
+
+~~~ CDDL
+problem-details = #6.65535(problem-details-map)
+problem-details-map = {
+  ? &(title: -1) => text
+  ? &(detail: -2) => text
+  ? &(instance: -3) => ~uri
+  standard-problem-detail-entries
+  custom-problem-detail-entries
+}
+standard-problem-detail-entries = (
+  * nint => any
+)
+custom-problem-detail-entries = (
+  * (uint/detail-label) => any
+)
+detail-label = text .regexp "[^:]+" / ~uri
+~~~
+{: #cddl title="Problem Detail Data Item"}
+
+The problem type is expressed in the tag number shown here as 65535.
+A tag has been registered as a generic problem type (see {{iana-tag}}),
+further tags for problem types can be registered (see {{sec-new-attributes}}).
 
-* "type" (id) - The problem type.  This is a mandatory element.
-* "title" (text) - A short, human-readable summary of the problem type.  It SHOULD NOT change from occurrence to occurrence of the problem.
-* "detail" (text) - A human-readable explanation specific to this occurrence of the problem.
-* "instance" (uri) - A URI reference that identifies the specific occurrence of the problem.  It may or may not yield further information if dereferenced.
+A number of problem details are predefined (more predefined details
+can be registered, see {{iana-spdk}}):
 
-Consumers MUST use "type" as primary identifiers for the problem type; the "title" string is advisory and included only for consumers who are not aware of the semantics of the "type" value.  An initial set of "type"s and corresponding "title"s is defined in {{sec-response-code-types}}.
+{:vspace}
+The title (key -1):
+: A short, human-readable summary of the problem type.  It SHOULD NOT change from occurrence to occurrence of the problem.
+
+The detail (key -2):
+: A human-readable explanation specific to this occurrence of the problem.
+
+The instance (key -3):
+: A URI reference that identifies the specific occurrence of the problem.  It may or may not yield further information if dereferenced.
+
+Consumers MUST use the type (tag number) as primary identifiers for the problem type; the "title" string is advisory and included only for consumers who are not aware of the semantics of the "type" value.
```

Comment:
Not sure about this strategy, it seems to lead to a proliferation of tag registrations.  I think sub-typing should be done by the `title` attribute.  We could register a certain URI to be the base for standard extensions that go through the IANA path, and leave the rest of the URI space open to custom extensions.

Reply:
s/title/type/

Reply:
Well, not sub-typing but typing is done through `type`, only that the indication is now coming from a tag rather than a field.

Actually, applying multiple tags would allow to have a multi-type problem details item, including the custom-problem-detail entry corresponding to each tag different than the one for the predefined Generic Problem Details type.

Reply:
I don't understand this idea.  Can you provide an example?  BTW, 7807bis is moving away from trying to handle multi-type problems: https://github.com/ietf-wg-httpapi/rfc7807bis/pull/37/files